File: solutions/cpp/0152.cpp
class Solution {
 public:
  int maxProduct(vector<int>& nums) {
    int ans = nums[0];
    int currMin = nums[0];
    int currMax = nums[0];

    for (int i = 1; i < nums.size(); ++i) {
      const int a = currMin * nums[i];
      const int b = currMax * nums[i];
      currMin = min({a, b, nums[i]});
      currMax = max({a, b, nums[i]});
      ans = max(ans, currMax);
    }

    return ans;
  }
};

File: solutions/java/0152.java
class Solution {
  public int maxProduct(int[] nums) {
    int ans = nums[0];
    int prevMin = nums[0];
    int prevMax = nums[0];

    for (int i = 1; i < nums.length; ++i) {
      int min = prevMin * nums[i];
      int max = prevMax * nums[i];
      prevMin = Math.min(nums[i], Math.min(min, max));
      prevMax = Math.max(nums[i], Math.max(min, max));
      ans = Math.max(ans, prevMax);
    }

    return ans;
  }
}